pkg/cmd/util/clientcmd/factory.go

```diff
@@ -45,12 +48,68 @@ func New(flags *pflag.FlagSet) *Factory {
 
 	// TODO: there should be two client configs, one for OpenShift, and one for Kubernetes
 	clientConfig := DefaultClientConfig(flags)
+	clientConfig = defaultingClientConfig{clientConfig}
 	f := NewFactory(clientConfig)
 	f.BindFlags(flags)
 
 	return f
 }
 
+// defaultingClientConfig detects whether the provided config is the default, and if
+// so returns an error that indicates the user should set up their config.
+type defaultingClientConfig struct {
+	nested kclientcmd.ClientConfig
+}
+
+// RawConfig calls the nested method
+func (c defaultingClientConfig) RawConfig() (kclientcmdapi.Config, error) {
+	return c.nested.RawConfig()
+}
+
+// Namespace calls the nested method, and if an empty config error is returned
+// it checks for the same default as kubectl - the value of POD_NAMESPACE or
+// "default".
+func (c defaultingClientConfig) Namespace() (string, bool, error) {
+	namespace, ok, err := c.nested.Namespace()
+	if err == nil {
+		return namespace, ok, nil
+	}
+	if !kclientcmd.IsEmptyConfig(err) {
+		return "", false, err
+	}
+	// TODO: can we inject the namespace as a file in the secret?
+	namespace = os.Getenv("POD_NAMESPACE")
+	if len(namespace) == 0 {
+		return api.NamespaceDefault, false, nil
+	}
+	return namespace, true, nil
+}
+
+// ClientConfig returns a complete client config
+func (c defaultingClientConfig) ClientConfig() (*kclient.Config, error) {
+	cfg, err := c.nested.ClientConfig()
+	if err == nil {
+		return cfg, nil
+	}
+
+	if !kclientcmd.IsEmptyConfig(err) {
+		return nil, err
+	}
+
+	if icc, err := kclient.InClusterConfig(); err == nil {
```

Comment:
Added todo, waiting to see if this goes green before pushing (then i'll stick merge on it unless you find more things)